File: scrimage-benchmarks/src/jmh/java/com/sksamuel/scrimage/benchmarks/ScaleBenchmarks.java
package com.sksamuel.scrimage.benchmarks;

import com.sksamuel.scrimage.ImmutableImage;
import com.sksamuel.scrimage.ScaleMethod;
import net.coobird.thumbnailator.Thumbnails;
import org.openjdk.jmh.annotations.*;
import org.openjdk.jmh.infra.Blackhole;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

@State(Scope.Benchmark)
public class ScaleBenchmarks {

   @Benchmark
   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.MILLISECONDS)
   public void testScaleBicubicScrimage(Blackhole blackhole) throws IOException {
      BufferedImage resized = ImmutableImage.loader().fromResource("/bench_colosseum.jpg")
         .scaleTo(600, 400, ScaleMethod.Bicubic)
         .awt();
      blackhole.consume(resized);
   }

   @Benchmark
   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.MILLISECONDS)
   public void testScaleProgressiveBilinearScrimage(Blackhole blackhole) throws IOException {
      BufferedImage resized = ImmutableImage.loader().fromResource("/bench_colosseum.jpg")
         .scaleTo(600, 400, ScaleMethod.Progressive)
         .awt();
      blackhole.consume(resized);
   }

   @Benchmark
   @BenchmarkMode(Mode.AverageTime)
   @OutputTimeUnit(TimeUnit.MILLISECONDS)
   public void testScaleProgressiveBilinearThumbnailator(Blackhole blackhole) throws IOException {
      BufferedImage resized = Thumbnails.of(getClass().getResourceAsStream("/bench_colosseum.jpg"))
         .size(600, 400)
         .keepAspectRatio(false)
         .asBufferedImage();
      blackhole.consume(resized);
   }
}

File: scrimage-benchmarks/src/main/kotlin/com/sksamuel/scrimage/benchmarks/ScaleBenchmarks.kt
package com.sksamuel.scrimage.benchmarks

import com.sksamuel.scrimage.ImmutableImage
import net.coobird.thumbnailator.Thumbnails
import java.awt.image.BufferedImage
import java.io.File
import javax.imageio.ImageIO
import kotlin.time.ExperimentalTime
import kotlin.time.measureTime

private val images = listOf(
   "/bears.jpg",
   "/elephant.jpg",
   "/falcon.jpg"
//   "/fox.jpg",
//   "/heron.jpg",
//   "/leopard.jpg",
//   "/lion.jpg",
//   "/peacock.jpg",
//   "/rhinoj.jpg"
)

private val libraries = listOf(
   Thumbnailator, Scrimage
)

@UseExperimental(ExperimentalTime::class)
fun main() {
   for (library in libraries) {
      println("Starting library ${library.name}")
      val duration = measureTime {
         for (image in images) {
            val scaled = library.scale(image, 600, 400)
            ImageIO.write(scaled, "JPEG", File(library.name + "_" + image.removePrefix("/")))
         }
      }
      println("Duration: ${duration.inMilliseconds}ms")
   }
}

interface Scale {
   fun scale(resource: String, width: Int, height: Int): BufferedImage
   val name: String
      get() = javaClass.simpleName
}

object Thumbnailator : Scale {
   override fun scale(resource: String, width: Int, height: Int): BufferedImage {
      return Thumbnails.of(javaClass.getResourceAsStream(resource))
         .size(width, height)
         .keepAspectRatio(false)
         .asBufferedImage()
   }
}

object Scrimage : Scale {
   override fun scale(resource: String, width: Int, height: Int): BufferedImage {
      return ImmutableImage.loader()
         .detectMetadata(false)
         .detectOrientation(false)
         .fromResource(resource)
         .scaleTo(width, height).awt()
   }
}
